source/GameInterface.Tests/Serialization/SerializerTests/WorkshopSerializationTest.cs

```diff
@@ -0,0 +1,116 @@
+        [Fact]
+        public void Workshop_Full_Serialization()
+        {
+            Settlement settlement = (Settlement)FormatterServices.GetUninitializedObject(typeof(Settlement));
+            MobileParty party = (MobileParty)FormatterServices.GetUninitializedObject(typeof(MobileParty));
+            Hero hero = (Hero)FormatterServices.GetUninitializedObject(typeof(Hero));
+
+            settlement.StringId = "mySettlement";
+            party.StringId = "myParty";
+            hero.StringId = "myHero";
+
+            MBObjectManager.Instance.RegisterObject(settlement);
+            MBObjectManager.Instance.RegisterObject(party);
+            MBObjectManager.Instance.RegisterObject(hero);
+
+            Workshop Workshop = new Workshop(settlement, "ws");
+
+            Capital.SetValue(Workshop, 5);
+            ConstructionTimeRemained.SetValue(Workshop, 100);
+            InitialCapital.SetValue(Workshop, 3);
+            InsideParty.SetValue(Workshop, party);
+            Level.SetValue(Workshop, 2);
+            NotRunnedDays.SetValue(Workshop, 566);
+            Upgradable.SetValue(Workshop, true);
+            _customName.SetValue(Workshop, new TextObject("I have a custom name"));
+            _owner.SetValue(Workshop, hero);
+            _productionProgress.SetValue(Workshop, new float[] { 0.6f, 0.7f });
+
+            BinaryPackageFactory factory = new BinaryPackageFactory();
+            WorkshopBinaryPackage package = new WorkshopBinaryPackage(Workshop, factory);
+
+            package.Pack();
+
+            byte[] bytes = BinaryFormatterSerializer.Serialize(package);
+
+            Assert.NotEmpty(bytes);
+
+            object obj = BinaryFormatterSerializer.Deserialize(bytes);
+
+            Assert.IsType<WorkshopBinaryPackage>(obj);
+
+            WorkshopBinaryPackage returnedPackage = (WorkshopBinaryPackage)obj;
+
+            Workshop newWorkshop = returnedPackage.Unpack<Workshop>();
+
+            Assert.Equal(Workshop.Capital, newWorkshop.Capital);
+            Assert.Equal(Workshop.ConstructionTimeRemained, newWorkshop.ConstructionTimeRemained);
+            Assert.Equal(Workshop.InitialCapital, newWorkshop.InitialCapital);
+            Assert.Equal(Workshop.Level, newWorkshop.Level);
+            Assert.Equal(Workshop.NotRunnedDays, newWorkshop.NotRunnedDays);
+            Assert.Equal(Workshop.Upgradable, newWorkshop.Upgradable);
+            Assert.Equal(_productionProgress.GetValue(Workshop), _productionProgress.GetValue(newWorkshop));
+            Assert.Equal(_settlement.GetValue(Workshop), _settlement.GetValue(newWorkshop));
+            Assert.Equal(_tag.GetValue(Workshop), _tag.GetValue(newWorkshop));
+
+            Assert.Equal(_customName.GetValue(Workshop).ToString(), _customName.GetValue(newWorkshop).ToString());
+
+            Assert.Equal(Workshop.InsideParty, newWorkshop.InsideParty);
```

Comment:
Fails but I'm pretty sure that's intentional because of MobileParty not being serialized yet?

Reply:
Correct, these will linger until the end of serialization